rsocket-core/src/main/java/io/rsocket/metadata/MimeMetadataCodec.java

```diff
@@ -0,0 +1,145 @@
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.util.CharsetUtil;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A flyweight class that can be used to encode/decode stream data mime types.
+ *
+ * <p>For more on the above metadata formats, see the corresponding <a
+ * href="https://github.com/rsocket/rsocket/blob/master/Extensions/PerStreamDataMimeTypesDefinition.md">Per
+ * stream data mime types definition</a>
+ */
+public class MimeMetadataCodec {
+  static final int STREAM_METADATA_KNOWN_MASK = 0x80; // 1000 0000
+  static final byte STREAM_METADATA_LENGTH_MASK = 0x7F; // 0111 1111
+
+  private MimeMetadataCodec() {}
+
+  /**
+   * made up of a single byte: this represents an encoded mime id, which can be further decoded
+   * using {@link #decode(ByteBuf)}
+   */
+  public static ByteBuf encodeWellKnowMime(ByteBufAllocator allocator, WellKnownMimeType mimeType) {
+    return allocator.buffer(1, 1).writeByte(mimeType.getIdentifier() | STREAM_METADATA_KNOWN_MASK);
+  }
+
+  /**
+   * Encode {@link WellKnownMimeType} or custom mime type into a newly allocated {@link ByteBuf}
+   *
+   * @param allocator the {@link ByteBufAllocator} to use to create the buffer.
+   * @param mimeType mime type
+   * @return the encoded mime
+   */
+  public static ByteBuf encode(ByteBufAllocator allocator, String mimeType) {
+    if (mimeType == null || mimeType.length() == 0) {
+      throw new IllegalArgumentException("Mime type null or length is zero");
+    }
+    WellKnownMimeType wkn = WellKnownMimeType.fromString(mimeType);
+    if (wkn == WellKnownMimeType.UNPARSEABLE_MIME_TYPE) {
+      return encodeCustomMime(allocator, mimeType);
+    } else {
+      return encodeWellKnowMime(allocator, wkn);
+    }
+  }
+
+  /**
+   * Encode multiple {@link WellKnownMimeType} or custom mime type into a newly allocated {@link
+   * ByteBuf}
+   *
+   * @param allocator the {@link ByteBufAllocator} to use to create the buffer.
+   * @param mimeTypes mime types
+   * @return the encoded mimes
+   */
+  public static ByteBuf encode(ByteBufAllocator allocator, List<String> mimeTypes) {
+    if (mimeTypes == null || mimeTypes.size() == 0) {
+      throw new IllegalArgumentException("Mime types empty");
+    }
+    CompositeByteBuf compositeMimeByteBuf = allocator.compositeBuffer();
+    for (String mimeType : mimeTypes) {
+      compositeMimeByteBuf.addComponents(true, encode(allocator, mimeType));
+    }
+    return compositeMimeByteBuf;
+  }
+
+  /**
+   * Encode a custom mime type into a newly allocated {@link ByteBuf}.
+   *
+   * <p>This larger representation encodes the mime type representation's length on a single byte,
+   * then the representation itself
+   *
+   * @param allocator the {@link ByteBufAllocator} to use to create the buffer.
+   * @param customMime a custom mime type to encode.
+   * @return the encoded mime
+   */
+  public static ByteBuf encodeCustomMime(ByteBufAllocator allocator, String customMime) {
```

Comment:
This does not need to be a public method. It's the same signature as `encode(ByteBufAllocator, String)` and provides the same end result.